src/do_command.c

```diff
@@ -577,7 +577,7 @@ static int child_process(entry * e, char **jobenv) {
 			if (mail && e->flags & MAIL_WHEN_ERR) {
 				int jobstatus = -1;
 				if (jobpid > 0) {
-					while (waitpid(jobpid, &jobstatus, WNOHANG) == -1) {
+					while (waitpid(jobpid, &jobstatus, 0) == -1) {
```

Comment:
This means that if the child process hangs indefinitely we will also keep the crond child waiting for it indefinitely. I am not sure about that. Perhaps we should iterate here with WNOHANG and sleeping for some time but not forever?

Reply:
I didn't want to divert from NetBSD and OpenBSD implementations, as they are using their implementation for 5 years, without issues.
In between these two, you can basically choose from not having email sent if the job hangs for a bit and then crash, or hanging until the job crashes.

I agree, that the best solution would be to wait for some time, then check the process and decide what to do with it. But how long we should wait?

Reply:
Actually given that we have SIGCHLD handler enabled we could just add a sleep() call before this waitpid() call with WNOHANG. That sleep will be interrupted by the SIGCHLD signal so there will be no extra waiting added.

I'd say the sleep() should be for 60 seconds or so, i.e. if the job finishes by that time it will send an e-mail in case of failure, otherwise not.